### PR TITLE
feat: Custom user agent enhancements merge to main

### PR DESCRIPTION
**Pull request by stocaaro**
#### Description of changes
This is a merge back to `main` of changes that have been reviewed as they where merged into `feat/user-agent-enhancements/main`. Changes are available for testing on the `v5-custom-user-agent` tag [in NPM](https://www.npmjs.com/package/aws-amplify/v/5.0.22-v5-custom-user-agent.f73abd2.28).

Feature work:
- https://github.com/aws-amplify/amplify-js/pull/11121)
- https://github.com/aws-amplify/amplify-js/pull/11324
- https://github.com/aws-amplify/amplify-js/pull/11384
- https://github.com/aws-amplify/amplify-js/pull/11410
- https://github.com/aws-amplify/amplify-js/pull/11425
- https://github.com/aws-amplify/amplify-js/pull/11437
- https://github.com/aws-amplify/amplify-js/pull/11442
- https://github.com/aws-amplify/amplify-js/pull/11459
- https://github.com/aws-amplify/amplify-js/pull/11452
- https://github.com/aws-amplify/amplify-js/pull/11457
- https://github.com/aws-amplify/amplify-js/pull/11467
- https://github.com/aws-amplify/amplify-js/pull/11468
- https://github.com/aws-amplify/amplify-js/pull/11465

Administrative/Unblocking
- https://github.com/aws-amplify/amplify-js/pull/11394
- https://github.com/aws-amplify/amplify-js/pull/11398
- https://github.com/aws-amplify/amplify-js/pull/11399
- https://github.com/aws-amplify/amplify-js/pull/11444
- https://github.com/aws-amplify/amplify-js/pull/11454

Merges from main
- https://github.com/aws-amplify/amplify-js/pull/11279
- https://github.com/aws-amplify/amplify-js/pull/11420
- https://github.com/aws-amplify/amplify-js/pull/11426
- https://github.com/aws-amplify/amplify-js/pull/11451
- https://github.com/aws-amplify/amplify-js/pull/11475

#### Description of how you validated changes
- Unit tests added/modified
- Integration steps running and working before preid releases
- We are working through manually testing of the preid release candidate to verify the user agent enhancements are working and not introducing issues

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amazon-cognito-identity-js/index.d.ts [hunk 1]
```diff
@@ -2,7 +2,11 @@ declare module 'amazon-cognito-identity-js' {
 	//import * as AWS from "aws-sdk";
 
 	export type NodeCallback<E, T> = (err?: E, result?: T) => void;
-	export type UpdateAttributesNodeCallback<E, T, K> = (err?: E, result?: T, details?: K) => void;
+	export type UpdateAttributesNodeCallback<E, T, K> = (
+		err?: E,
+		result?: T,
+		details?: K
+	) => void;
```
Comment (AllanZhengYP) [inline packages/amazon-cognito-identity-js/index.d.ts:9]: super nit: remove this diff?

File: packages/amazon-cognito-identity-js/internals/package.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+	"name": "amazon-cognito-identity-js/internals",
+	"types": "./index.d.ts",
+	"main": "../lib/internals/index.js",
+	"module": "../es/internals/index.js",
+	"react-native": "../es/internals/index.js",
+	"sideEffects": false
+}
```
Comment (AllanZhengYP) [inline packages/amazon-cognito-identity-js/internals/package.json:8]: nit: trailing new line

File: packages/amazon-cognito-identity-js/src/Platform/constants.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export const Framework = {
+	None: '0',
+	ReactNative: '1',
+};
+
+export const authCategory = 'auth';
```
Comment (AllanZhengYP) [inline packages/amazon-cognito-identity-js/src/Platform/constants.js:6]: nit: All cap for string constants?

File: packages/analytics/__tests__/Providers/AWSPinpointProvider.test.ts [hunk 4]
```diff
@@ -285,7 +292,12 @@ describe('AnalyticsProvider test', () => {
 				const params = { event: { name: 'custom event', immediate: true } };
 				await analytics.record(params, { resolve, reject });
 				expect(mockPutEvents).toBeCalledWith(
-					{ credentials, region: 'region' },
+					expect.objectContaining({
+						credentials,
+						region: 'region',
+						userAgentValue: expect.stringMatching(expectedRecordUserAgentRegex),
+						// userAgentValue: expect.stringMatching(new RegExp('^aws-amplify')),
```
Comment (AllanZhengYP) [inline packages/analytics/__tests__/Providers/AWSPinpointProvider.test.ts:299]: Nit: clean comments?

File: packages/amazon-cognito-identity-js/internals/package.json (hunk 2 above)
Comment (AllanZhengYP) [reply packages/amazon-cognito-identity-js/internals/package.json:8]: Also make sure files in internals folder added to the `files` in the package.json at the package root.

File: packages/api-graphql/package.json [hunk 5]
```diff
@@ -64,7 +65,7 @@
       "name": "API (GraphQL client)",
       "path": "./lib-esm/index.js",
       "import": "{ Amplify, GraphQLAPI }",
-      "limit": "87.1 kB"
+      "limit": "88.5 kB"
```
Comment (AllanZhengYP) [inline packages/api-graphql/package.json:68]: Very nit: why did this increased 1.4 kB whereas other packages increased 1.2 kB?

File: packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts [hunk 6]
```diff
@@ -359,7 +370,10 @@ export class AmazonAIIdentifyPredictionsProvider extends AbstractIdentifyPredict
 		this.rekognitionClient = new RekognitionClient({
 			region,
 			credentials,
-			customUserAgent: getAmplifyUserAgent(),
+			customUserAgent: getAmplifyUserAgent({
+				category: Category.Predictions,
+				action: PredictionsAction.Identify,
+			}),
```
Comment (AllanZhengYP) [inline packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts:376]: I see this repeats a few times, can we set it to a variable and reuse it?

**Review by AllanZhengYP — DISMISSED** [review]
I skimmed through it since most of the changes were reviewed already, and left some questions. None of them are blocking. It looks good!

File: .circleci/config.yml [hunk 7]
```diff
@@ -2060,15 +2061,15 @@ workflows:
...
-              browser: [chrome]
+      #- integ_javascript_geo:
+      #    requires:
+      #      - integ_setup
+      #      - build
+      #    filters:
+      #      <<: *releasable_branches
+      #    matrix:
+      #      parameters:
+      #        browser: [chrome]
```
[the remark this answers is not in the document]
Comment (stocaaro) [reply .circleci/config.yml:2072]: It depends on static build artifacts that aren't updated by the build. Not sure what we're getting from this re: integ testing changes.

File: packages/analytics/__tests__/Providers/AWSPinpointProvider.test.ts (hunk 4 above)
Comment (stocaaro) [reply packages/analytics/__tests__/Providers/AWSPinpointProvider.test.ts:299]: Removed

File: packages/amazon-cognito-identity-js/src/Platform/constants.js (hunk 3 above)
Comment (stocaaro) [reply packages/amazon-cognito-identity-js/src/Platform/constants.js:6]: Sure

File: packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts (hunk 6 above)
Comment (stocaaro) [reply packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts]: Agreed. Factored out into a method for reuse.

File: packages/amazon-cognito-identity-js/index.d.ts (hunk 1 above)
Comment (stocaaro) [reply packages/amazon-cognito-identity-js/index.d.ts:9]: Removed

**Review by AllanZhengYP — APPROVED** [review]

**Review by elorzafe — DISMISSED** [review]
Approved with minor non blocking comments

**Review by jimblanc — DISMISSED** [review]
Gave it a pass paying special attention to configuration & package interfaces, LGTM! Great work to you and Erin on this 🥳

**Review by jimblanc — APPROVED** [review]

File: .circleci/config.yml [hunk 8]
```diff
@@ -2060,15 +2060,15 @@ workflows:
...
-      - integ_javascript_geo:
-          requires:
-            - integ_setup
-            - build
-          filters:
-            <<: *releasable_branches
-          matrix:
-            parameters:
-              browser: [chrome]
+      #- integ_javascript_geo:
```
Comment (stocaaro) [inline .circleci/config.yml:2063]: We will take a loot at turning this test on again in `main`. Theres a usage of static assets here that I wasn't able to figure out how to fix to make this test value add. Hopefully I'm missing something. Will validate with Francisco before releasing anything.

**Review by AllanZhengYP — APPROVED** [review]

**Review by iartemiev — APPROVED** [review]
Looked over the `api*/*` and `datastore/*` changes - no surprises vs. already-approved PRs. LGTM 👍🏻

**Pull request merged**